src/proc/constant_evaluator.rs

```diff
@@ -62,169 +82,107 @@ pub enum ConstantSolvingError {
-        self.solve_impl(expr, ExprType::Regular, true)
+    ) -> Result<Handle<Expression>, ConstantEvaluatorError> {
+        match self.expressions[expr] {
+            Expression::Literal(_)
+            | Expression::ZeroValue(_)
+            | Expression::Compose { .. }
+            | Expression::Splat { .. } => Ok(expr),
+            Expression::Constant(c) => {
+                if let Some(const_expressions) = self.const_expressions {
+                    self.copy_from(self.constants[c].init, const_expressions)
+                } else {
+                    Ok(self.constants[c].init)
+                }
+            }
+            _ => Err(ConstantEvaluatorError::SubexpressionsAreNotConstant),
+        }
     }
 
-    pub fn solve_impl(
+    pub fn try_eval_and_append(
         &mut self,
-        expr: Handle<Expression>,
-        expr_type: ExprType,
-        top_level: bool,
-    ) -> Result<Handle<Expression>, ConstantSolvingError> {
-        let expressions = match expr_type {
-            ExprType::Regular => self.expressions,
-            ExprType::Constant => self.const_expressions,
-        };
-        let span = expressions.get_span(expr);
-        match expressions[expr] {
-            ref expression @ (Expression::Literal(_) | Expression::ZeroValue(_)) => match expr_type
-            {
-                ExprType::Regular => Ok(self.register_constant(expression.clone(), span)),
-                ExprType::Constant => Ok(expr),
-            },
-            Expression::Compose { ty, ref components } => match expr_type {
-                ExprType::Regular => {
-                    let mut components = components.clone();
-                    for component in &mut components {
-                        *component = self.solve_impl(*component, expr_type, false)?;
-                    }
-                    Ok(self.register_constant(Expression::Compose { ty, components }, span))
-                }
-                ExprType::Constant => Ok(expr),
-            },
-            Expression::Constant(constant) => {
-                if top_level {
-                    match expr_type {
-                        ExprType::Regular => {
-                            Ok(self.register_constant(Expression::Constant(constant), span))
-                        }
-                        ExprType::Constant => Ok(expr),
-                    }
-                } else {
-                    self.solve_impl(self.constants[constant].init, ExprType::Constant, false)
+        expr: &Expression,
+        span: Span,
+    ) -> Result<Handle<Expression>, ConstantEvaluatorError> {
+        match *expr {
+            Expression::Literal(_) | Expression::ZeroValue(_) | Expression::Constant(_) => {
+                Ok(self.register_constant(expr.clone(), span))
+            }
+            Expression::Compose { ref components, .. } => {
+                for component in components {
+                    self.check_and_get(*component)?;
```

Comment:
Since `check_and_get` can return a different handle, shouldn't this always be rebuilding the component vector?

Reply:
Marking as resolved as we've taken the discussion to: https://github.com/gfx-rs/naga/pull/2493.